core/shared/src/main/scala/org/http4s/Message.scala

```diff
@@ -490,6 +458,38 @@ final class Request[F[_]] private (
 
 object Request {
 
+  /** A [[Request]] that doesn't have a body requiring effectful evaluation.
+    *  A [[Request.Pure]] is always an instance of [[Request[F]]], regardless of `F`.
+    */
+  type Pure = Request[fs2.Pure]
+
+  implicit final class InvariantOps[F[_]](private val self: Request[F]) extends AnyVal {
```

Comment:
What about the methods in here? 😅

Reply:
These ones could be done as partial application I think. I'm not too bothered either way but maybe we should do that as it's a more familiar pattern.

Reply:
I have this a try, but the problem is that the helper classes can’t extend `AnyVal` because they would be inner classes - and if they’re defined outside the class they would need to have an `F` parameter, which would have to occur in contravariant position at the call site where they’re instantiated, which wouldn’t compile…

let’s just leave it as-is?

Reply:
Thanks! I gave it a try myself, this is what I came up with. Let me know what you think :)

- https://github.com/bplommer/http4s/pull/1

Reply:
I'm not wild about that change. Feels like it degrades the user experience for the sake of making implementation slightly simpler. I'm hilling to be outvoted on this though if it's what other people prefer.

Kudos for using git the way it's supposed to be used by opening a PR against my PR though

Reply:
> Feels like it degrades the user experience for the sake of making implementation slightly simpler.

Well, there's trade-offs. IMO it makes the UX better in terms of discoverability. When those methods are hidden on implicit classes they are trickier to find.

Reply:
That’s a fair point. We could try your change and see if people complain 😅